rope.c

```diff
@@ -131,6 +131,85 @@ size_t rope_write_cstr(rope *r, uint8_t *dest) {
   return num_bytes + 1;
 }
 
+static inline uint8_t *skip_utf8(uint8_t *str, size_t chars, size_t bytelen);
+
+static inline size_t copy_utf8(uint8_t *dst, const uint8_t *src, size_t *chars,
+                               size_t max_size);
+
+size_t rope_write_substr(rope *rope, uint8_t *buf, size_t *bufsize,
+                         size_t from, size_t to) {
+  size_t len = rope_char_count(rope);
+  if (to > len) {
+    to = len;
+  }
+
+  rope_node *node = &rope->head;
+  size_t skipped = 0;
+
+  // Find the starting node.
+  while (skipped + node->nexts[0].skip_size < from) {
+    // Use the skip list.
+    uint8_t height = node->height;
+    uint8_t i;
+    for (i = 1; i < height; ++i) {
+      if (skipped + node->nexts[i].skip_size >= from) {
+        // Too far. Look at the next node's skip list.
+        break;
```

Comment:
Shouldn't this be `continue`?

Reply:
This breaks the for loop, which continues the outer while loop, but I'll replace this with `iter_at_char_pos`.